constraint-solver/src/solver/linearized.rs

```diff
@@ -0,0 +1,449 @@
+        self.solver.add_algebraic_constraints(constraints);
+    }
+
+    fn add_bus_interactions(
+        &mut self,
+        bus_interactions: impl IntoIterator<Item = BusInteraction<GroupedExpression<T, Variable<V>>>>,
+    ) {
+        let bus_interactions = bus_interactions
+            .into_iter()
+            .map(|bus_interaction| {
+                bus_interaction
+                    .fields()
+                    .map(|expr| {
+                        self.linearizer
+                            .substitute_by_var(expr.clone(), &mut || self.var_dispenser.next_var())
+                    })
+                    .collect::<BusInteraction<_>>()
+            })
+            .collect_vec();
+        // The substitution was not yet linearized.
+        let to_add = std::mem::take(&mut self.linearizer.constraints_to_add);
+        self.add_algebraic_constraints(to_add);
+        self.solver.add_bus_interactions(bus_interactions);
+    }
+
+    fn add_range_constraint(
+        &mut self,
+        variable: &Variable<V>,
+        constraint: RangeConstraint<T::FieldType>,
+    ) {
+        self.solver.add_range_constraint(variable, constraint);
+    }
+
+    fn retain_variables(&mut self, variables_to_keep: &HashSet<Variable<V>>) {
+        // There are constraints that only contain `Variable::Linearized` that
+        // connect quadratic terms with the original constraints. We could try to find
+        // those, but let's just keep all of them for now.
+        let mut variables_to_keep = variables_to_keep.clone();
+        variables_to_keep.extend(self.var_dispenser.all_dispensed_vars());
+        self.solver.retain_variables(&variables_to_keep);
+    }
+
+    fn range_constraint_for_expression(
+        &self,
+        expr: &GroupedExpression<T, Variable<V>>,
+    ) -> RangeConstraint<T::FieldType> {
+        // Ask the solver directly for the range constraint of the expression.
+        let direct = self.solver.range_constraint_for_expression(expr);
+        // See if we have a direct substitution for the expression by a variable.
+        let simple_substituted = self
+            .linearizer
+            .try_substitute_by_existing_var(expr)
+            .map(|expr| self.solver.range_constraint_for_expression(&expr))
+            .unwrap_or_default();
+        // Try to re-do the linearization
+        let substituted = self
+            .linearizer
+            .try_linearize_existing(expr.clone())
+            .map(|expr| self.solver.range_constraint_for_expression(&expr))
+            .unwrap_or_default();
+        direct
+            .conjunction(&simple_substituted)
+            .conjunction(&substituted)
+    }
+
+    fn are_expressions_known_to_be_different(
+        &mut self,
+        a: &GroupedExpression<T, Variable<V>>,
+        b: &GroupedExpression<T, Variable<V>>,
+    ) -> bool {
+        let a = iter::once(a.clone()).chain(
+            (!a.is_affine())
```

Comment:
In analogy to the function above, what about `self.linearizer.try_substitute_by_existing_var(...)`?

Reply:
In `range_constraint_for_expression`, we want to have a single variable because range constraints are stored for variables. But here, we actually prefer complex expressions over single variables because it is more likely that the difference cancels out. This would also apply for single variables that are the same, but then the expressions should already be the same and we gain nothing by transforming it into a variable. I'll check again and maybe add it for simplicity.